modules/local/tasks.nf

```diff
@@ -0,0 +1,381 @@
+    
+    sample_id=${sample_id_gr}
+    
+    dname=${outdir}/\${sample_id}
+        
+    Rscript $projectDir/bin/characterization_SPOTlight.R \${dname}/ ${sample_info.st_data_dir}
+    echo "completed" > "output.out" && outpath=`pwd`/output.out
+    """
+}
+
+
+/* 
+ * Resolution enhancement and spatial clustering with BayesSpace
+ */
+ process CLUSTERING_WITH_BAYESSPACE {
+    
+    label "r_process"
+     
+    input:
+    val sample_state
+    val outdir
+    
+    output:
+    tuple env(sample_id), env(outpath)
+    
+    script:
+    def sample_id_gr = sample_state[0]
+    def fileName = String.format("%s/sample_%s.json", outdir, sample_id_gr)
+    sample_info = new JsonSlurper().parse(new File(fileName))
+        
+    """
+    #!/bin/bash
+    
+    sample_id=${sample_id_gr}
+    
+    dname=${outdir}/\${sample_id}
+    
+    Rscript $projectDir/bin/characterization_BayesSpace.R \${dname}/
+    echo "completed" > "output.out" && outpath=`pwd`/output.out
+    """
+}
+
+
+/* 
+ * SpatialDE
+ */
+ process ST_SPATIALDE {
+     
+    label "python_process"
+    
+    input:
+    val sample_state
+    val outdir
+    
+    output:
+    tuple env(sample_id), env(outpath)
+    
+    script:
+    def sample_id_gr = sample_state[0]
+    def fileName = String.format("%s/sample_%s.json", outdir, sample_id_gr)
+    sample_info = new JsonSlurper().parse(new File(fileName))
+        
+    """
+    #!/bin/bash
+    
+    sample_id=${sample_id_gr}
+    
+    dname=${outdir}/\${sample_id}
+       
+    python $projectDir/bin/stSpatialDE.py \${dname}/ st_adata_norm.h5ad
+    echo "completed" > "output.out" && outpath=`pwd`/output.out
+    """
+}
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+/* 
+ * Clustering etc.
+ */
+ process ST_CLUSTERING {
+     
+    label "python_process"
+    
+    input:
+    val sample_state
+    val outdir
+    
+    output:
+    tuple env(sample_id), env(outpath)
+    
+    script:
+    def sample_id_gr = sample_state[0]
+    def fileName = String.format("%s/sample_%s.json", outdir, sample_id_gr)
+    sample_info = new JsonSlurper().parse(new File(fileName))
+        
+    """
+    #!/bin/bash
+    
+    sample_id=${sample_id_gr}
+    
+    dname=${outdir}/\${sample_id}
+         
+    python $projectDir/bin/stClusteringWorkflow.py \${dname}/
+    echo "completed" > "output.out" && outpath=`pwd`/output.out
+    """
+}
+
+
+/* 
+ * Report
+ */
+ process ALL_REPORT {
+     
+    label "python_process_low"
+    
+    input:
+    val sample_state
+    val outdir
+    
+    output:
+    tuple env(sample_id), env(outpath)
+    
+    script:
+    def sample_id_gr = sample_state[0]
+    def fileName = String.format("%s/sample_%s.json", outdir, sample_id_gr)
+    sample_info = new JsonSlurper().parse(new File(fileName))
+        
+    """
+    #!/bin/bash
+    
+    sample_id=${sample_id_gr}
+    
+    dname=${outdir}/\${sample_id}
+    
+    echo \${dname}/  
+    echo "completed" > "output.out" && outpath=`pwd`/output.out
+    """
+}
+
```

Comment:
This contains all the processes for the different subworkflows, but should probably be divided up into individual modules?